addons/blocklydevtools/userscript.js

```diff
@@ -0,0 +1,419 @@
+                .replace(/\n/g, "<br>")
+                .replace(/ /g, "&nbsp;");
+        }
+    }
+    function getBlocklyWorkspace() { //Get the svg workspace
+        return document.querySelector("svg.blocklySvg>g.blocklyWorkspace");
+    }
+    function getBlocklyBlockCanvas() { //get the block canvas from the workspace
+        return getBlocklyWorkspace().querySelector("g.blocklyBlockCanvas");
+    }
+    function getSvgPathFromBlock(blockElem) { //Get the svg path from an svg block
+        if (isNewBlocklyBlock(blockElem)) {
+            return blockElem.querySelector("g.blocklyDraggable[data-id]:not(.blocklyInsertionMarker)>path.blocklyPath[d]");
+        }
+    }
+    function isNewBlocklyBlock(node) { //Is node a block svg that has not yet been hooked.
+        return node && node instanceof Element && node.tagName === "g" && node.classList.contains("blocklyDraggable") && !node.classList.contains("blocklyInsertionMarker") && node.hasAttribute("data-id") && !node.hasAttribute("data-blocklydev-hooked");
+    }
+    function makeBlockEditor(blockId) { //Make the XML editor for a blockId
+        function getXmlFromBlockId(id, dom) { //Utility function to get the block's XML string from a Blockly serialised DOM.
+            for (let i = 0; i < dom.children.length; i++) {
+                const element = dom.children[i];
+                if (element.tagName.toLowerCase() === "block" && element.getAttribute("id") === id) {
+                    return element.outerHTML;
+                }
+            }
+        }
+
+        var originalXml = getXmlFromBlockId(blockId, dom); //Store the block's original, un-edited XML.
+
+        var container = document.createElement("div"); //Create the editor
+
+        container.setAttribute("data-isblocklydeveditor", "true"); //Give it an identifying attribute
+        container.contentEditable = true; //Allow it's contents to be edited
+
+        //Styling
+        container.style.padding = "10px";
+        container.style.userSelect = "text";
+        container.style.cursor = "auto";
+        container.style.font = "12pt monospace";
+        container.style.boxShadow = "0px 0px 10px rgba(0,0,0,0.5)";
+        container.style.color = "yellowgreen";
+        container.style.border = "2px solid white";
+        container.style.zIndex = "998";
+        container.style.borderRadius = "0.8rem";
+        container.style.width = "max-content";
+        container.style.backgroundColor = "rgb(0,0,30)";
+        container.style.overflowX = "hidden";
+        container.style.maxHeight = "100vh";
+        container.style.overflowY = "scroll";
+        container.style.caretColor = "white";
+
+        //Disable spellcheck and autocomplete.
+        container.setAttribute("autocomplete", "false");
+        container.setAttribute("spellcheck", "false");
+
+        //Set the editor's contents to the formatted original XML.
+        container.innerHTML = formatXml(originalXml, true);
```

Comment:
Ideally all uses of `innerHTML` would be replaced with `innerText` or `textContext` where possible.

Reply:
Yes, I know, but I've found innerText and textContent to be very hard to work with (in terms of getting indentation to work). However, I think I've mitigated any XSS by replacing any `<` or `>` characters with `&lt;` and `&gt;` in `formatXml()`.

Reply:
I have found replaced some unnecessary usage of innerHTML with innerText when the 'icons' are set for the buttons.